core/drivers/crypto/crypto_api/acipher/ecc.c

```diff
@@ -340,10 +340,161 @@ static TEE_Result ecc_shared_secret(struct ecc_keypair *private_key,
 	return ret;
 }
 
+static TEE_Result ecc_sm2_encrypt(struct ecc_public_key *key,
+			      const uint8_t *src, size_t src_len,
+			      uint8_t *dst, size_t *dst_len)
+{
+	TEE_Result ret = TEE_ERROR_BAD_PARAMETERS;
+	struct drvcrypt_cipher_data cdata = { };
+	struct drvcrypt_ecc *ecc = NULL;
+	size_t ciphertext_len = 0;
+	size_t size_bytes = 0;
+	size_t c1_len = 0;
+
+	ecc = drvcrypt_get_ops(CRYPTO_ECC);
+
+	size_bytes = get_ecc_key_size_bytes(key->curve);
+	if (!size_bytes) {
+		CRYPTO_TRACE("Invalid size_bytes");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	/* Uncompressed form indicator */
+	dst[0] = 0x04;
+
+	c1_len = 2 * size_bytes;
+	ciphertext_len = c1_len + src_len + TEE_SM3_HASH_SIZE;
+
+	cdata.key = key;
+	cdata.size_sec = size_bytes;
+	cdata.plaintext.data = (uint8_t *)src;
+	cdata.plaintext.length = src_len;
+	cdata.ciphertext.data = dst + 1;
+	cdata.ciphertext.length = ciphertext_len;
+
+	ret = ecc->encrypt(&cdata);
+
+	/* Set the ciphertext length */
+	if (!ret || ret == TEE_ERROR_SHORT_BUFFER)
+		*dst_len = cdata.ciphertext.length + 1;
+
+	return ret;
+}
+
+static TEE_Result ecc_encrypt(struct ecc_public_key *key,
+			      const uint8_t *src, size_t src_len,
+			      uint8_t *dst, size_t *dst_len)
+{
+	struct drvcrypt_ecc *ecc = NULL;
+
+	if (!key || !src || !dst) {
+		CRYPTO_TRACE("Input parameters reference error");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	ecc = drvcrypt_get_ops(CRYPTO_ECC);
+	if (!ecc || !ecc->encrypt)
+		return TEE_ERROR_NOT_IMPLEMENTED;
+
+	switch (key->curve) {
+		case TEE_ECC_CURVE_SM2:
+			return ecc_sm2_encrypt(key, src, src_len, dst, dst_len);
+		default:
+			return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+}
+
+static TEE_Result ecc_sm2_decrypt(struct ecc_keypair *key,
+			      const uint8_t *src, size_t src_len,
+			      uint8_t *dst, size_t *dst_len)
+{
+	TEE_Result ret = TEE_ERROR_BAD_PARAMETERS;
+	struct drvcrypt_cipher_data cdata = { };
+	struct drvcrypt_ecc *ecc = NULL;
+	uint8_t *ciphertext = NULL;
+	size_t ciphertext_len = 0;
+	size_t size_bytes = 0;
+	size_t c1_len = 0;
+	size_t c2_len = 0;
+	/* Point Compression */
+	uint8_t pc = 0;
+
+
+	ecc = drvcrypt_get_ops(CRYPTO_ECC);
+
+	size_bytes = get_ecc_key_size_bytes(key->curve);
+	if (!size_bytes) {
+		CRYPTO_TRACE("Invalid size_bytes");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	pc = src[0];
+	switch (pc) {
+	case 0x02:
+	case 0x03:
+		/* Compressed form */
+		return TEE_ERROR_NOT_SUPPORTED;
+	case 0x04:
+		/* Uncompressed form */
+		ciphertext = (uint8_t *)src + 1;
+		ciphertext_len = src_len - 1;
+		c1_len = 2 * size_bytes;
+		break;
+	case 0x06:
+	case 0x07:
+		/* Hybrid form */
+		return TEE_ERROR_NOT_SUPPORTED;
+	default:
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (SUB_OVERFLOW(ciphertext_len, (c1_len + TEE_SM3_HASH_SIZE), &c2_len))
```

Comment:
can remove inner parentheses for arg#2.
maybe use a single `len` or `plain_len` variable to compute length, instead of `c1_len` and `c2_len`.